### PR TITLE
add Gradle build support, initial version (fixes #118)

### LINKED ISSUE #118: add support for Gradle builds in addition to Maven
Would there be any interest in a contribution to this project to build Gradle projects, in addition to Maven?

It doesn't seem that hard to [add a something like this here](https://github.com/fabric8io-images/s2i/blob/master/java/images/jboss/s2i/assemble):

    if [ -f "${S2I_SOURCE_DIR}/*.gradle*" ]; then

note the use of `*.gradle*` in order to also allow `build.gradle.kts` for the Gradle Kotlin DSL, and [myProjectName.gradle](https://docs.gradle.org/current/userguide/command_line_interface.html#_environment_options).

if present, then run something like `./gradlew --gradle-user-home=${S2I_ARTIFACTS_DIR}/gradle-home --project-cache-dir=${S2I_ARTIFACTS_DIR}/gradle-project-cache build` (TBC, probably with a GRADLE_ARGS and/or GRADLE_ARGS_APPEND, similar to the MAVEN_ARGS and MAVEN_ARGS_APPEND).

I would limit this to support only the Gradle wrapper, and not even bother to add a Gradle distribution into the image. Just because using the wrapper is VERY widespread in Gradle based project (contrary to [the maven-wrapper](https://github.com/takari/maven-wrapper)), and saves a ton of "which version" hassles, which for frequently releasing Gradle is even more important than with Maven (BTW surprised we're on ancient Maven 3.3.3 instead of latest 3.5.2 here...). The overhead of downloading a Gradle distribution IMHO is manageable with incremental builds by preserving the --gradle-user-home and --project-cache-dir accross builds; in the bigger picture of typically many many dependencies, the Gradle distribution itself is really just another such dependency. 

BTW incremental builds with Gradle should be blazingly fast, not only avoiding dependency re-downloads, but truly incrementallying building code changes themselves (which never really works reliably in Maven; BTW I'm surprised [we don't do clean package instead of just package here](https://github.com/fabric8io-images/s2i/blob/master/java/images/jboss/s2i/assemble)).

It should also include the equivalent of https://github.com/fabric8io-images/s2i/issues/110.

### PR DESCRIPTION
**Pull request by vorburger**

### CODE REVIEW

File: java/images/jboss/s2i/assemble [hunk 1]
```diff
@@ -47,6 +47,32 @@ function get_output_dir() {
...
+    if [ "${ARTIFACT_DIR:0:1}" = "/" ]; then
+       echo "ARTIFACT_DIR \"${ARTIFACT_DIR}\" must not be absolute but relative to the source directory"
+       exit 1
+    fi
+    dir="${S2I_SOURCE_DIR}/${ARTIFACT_DIR}"
+  fi
+
+  # Normalize dir
+  # TODO how do you get this to check not one but two (../..) up?
+#  dir=$(echo ${dir} | tr -s /)
```
Comment (rhuss) [inline java/images/jboss/s2i/assemble]: what is this section supposed to do ?

File: java/images/jboss/s2i/assemble [hunk 2]
```diff
@@ -111,6 +136,10 @@ function setup_maven() {
   fi
 }
 
+function setup_gradle() {
```
Comment (rhuss) [inline java/images/jboss/s2i/assemble]: Can we remove it for now if not used ?

File: java/images/jboss/s2i/assemble [hunk 3]
```diff
@@ -160,14 +189,59 @@ function build_maven() {
...
+  local build_dir=$1
+
+  # Where to put the artifacts
+  local app_dir=$2
+
+  # Default args
+  local gradle_args=${GRADLE_ARGS:-build -x test}
+
+  # If there is no user provided GRADLE_OPTS, use options from run-java.sh
+  if [ -f ${RUN_JAVA_DIR}/run-java.sh -a -z "${GRADLE_OPTS}" ]; then
```
Comment (rhuss) [inline java/images/jboss/s2i/assemble]: using `[ .... ] && [ .... ]` proved to be more robust to me. `-a` fails in certain edge case (which I dont remember right now ;-)

File: java/images/jboss/s2i/assemble [hunk 4]
```diff
@@ -160,14 +189,59 @@ function build_maven() {
...
+    export GRADLE_OPTS=$(${RUN_JAVA_DIR}/run-java.sh options)
+  fi
+
+  if [ ! -z "${GRADLE_OPTS}" ]; then
```
Comment (rhuss) [inline java/images/jboss/s2i/assemble]: `-n` ?

File: java/images/jboss/s2i/assemble [hunk 5]
```diff
@@ -160,14 +189,59 @@ function build_maven() {
...
+
+  # ==============
+  # Copy artifacts
+  echo "Copying Gradle artifacts from ${build_dir} to ${app_dir} ..."
+  copy_artifacts ${build_dir} ${app_dir}
+  check_error "copying artifacts from ${build_dir} to ${app_dir}" $?
+
+  # ======================
+  # TODO Remove repo if desired
+#  if [ "x${MAVEN_CLEAR_REPO}" != "x" ]; then
```
Comment (rhuss) [inline java/images/jboss/s2i/assemble]: Why commented out ?

**Comment by jamesnetherton** [comment]
If we are going to support Gradle then we should restrict its implementation to the community images initially.

**Comment by dhirajsb** [comment]
agree with @jamesnetherton

**Comment by vorburger** [comment]
@rhuss I'm hoping to be able to make time to pick this up after my other pending PRs are in; the idea would be to cover this using my new self testing approach for this example as well.

@jamesnetherton @dhirajsb sure OK for me initially community images only; so I'll put a big IFF into the respective templates - OK for you as well @rhuss ?

**Comment by rhuss** [comment]
@vorburger yes, sounds like a plan. I'm still a bit hesitant ans I'm not sure how good we can support the Gradle support in case of an error, but hey, let's try it ;-)

**Comment by vorburger** [comment]
> BTW, @vorburger I added you with write access to this repo so that you then can work on and fix Gradle issues ;-P

OK, cool - will do my best!

**Comment by maslick** [comment]
running s2i  with ``-e GRADLE_CLEAR_REPO=true`` has no effect on the image size:
```
s2i build https://github.com/maslick/see-i-see-d.git#jdk11 fabric8/s2i-java:latest-java11 vesna:jdk11 -e GRADLE_CLEAR_REPO=true
```

**Comment by rhuss** [comment]
@vorburger could you please have a look ?

**Comment by vorburger** [comment]
@maslick sorry I don't really have the time to help with this project anymore, but if you'd like to raise a PR putting what you are suggesting above into code, then perhaps someone will help to review and merge it.

**Pull request merged**